src/filtering/filterParser.ts

```diff
@@ -0,0 +1,421 @@
+   * @param off the offset to look at
+   * @returns a Token if it's within the range of the stream otherwise null
+   */
+  public peek(off: number): Token | null {
+    const n = this.pos + off;
+    if (n >= 0 && n < this.tokens.length) {
+      return this.tokens[this.pos + off];
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Returns the next token in the stream and moves the position forward.
+   * @param types an array of `TokenTypes` check the current token against
+   * @returns The next token in the stream or null if it's the end of the stream
+   * @throws {ParserError} if the token does not match the `types` param or if we attempt to read beyond the end of the stream
+   */
+  public consume(types?: TokenType[]): Token | null {
+    if (types) {
+      if (this.pos < this.tokens.length) {
+        if (types.includes(this.tokens[this.pos].type)) {
+          return this.tokens[this.pos++];
+        } else {
+          throw new ParserError(
+            `Expected token from types [${types}] at token ${
+              this.peek(0)?.value
+            }` // TODO: better error message?
+          );
+        }
+      } else {
+        throw new ParserError(`Missing token at end. Expected: ${types}`);
+      }
+    } else {
+      if (this.pos < this.tokens.length) {
+        return this.tokens[this.pos++];
+      } else if (this.pos === this.tokens.length) {
+        return null;
+      } else {
+        throw new ParserError('Attempt to read beyond end');
+      }
+    }
+  }
+}
+
+/**
+ * Class representing a predicate e.g. `CHANNEL_1 < 1` or `activeExperiment is not null`
+ *
+ * Predicate ::= channel COMPOP param | channel UNARYOP
+ */
+class Predicate {
+  param1: string | number | undefined;
+  param2: string | number | undefined;
+  op: Token | undefined;
+
+  /**
+   * @param input `Input` stream to use
+   * @throws {ParserError} if the predicate has invalid syntax e.g. number on LHS, no operator etc.
+   */
+  constructor(input: Input) {
+    let token = input.peek(0);
+    if (token !== null) {
+      if (token.type === 'channel') {
+        this.param1 = convertChannel(token.value);
+        input.consume();
+      } else if (token.type === 'string') {
+        throw new ParserError(
+          `Unexpected string on left hand side of expression: ${token.value}`
+        );
+      } else if (token.type === 'number') {
+        throw new ParserError(
+          `Unexpected number on left hand side of expression: ${token.value}`
+        );
+      } else {
+        throw new ParserError(`Unexpected: ${token.value}`);
+      }
+    } else {
+      throw new ParserError('Missing operand');
+    }
+    this.op = input.consume(['compop', 'unaryop']) as Token;
+    if (this.op.type === 'compop') {
+      token = input.peek(0);
+      if (token !== null) {
+        if (token.type === 'channel') {
+          this.param2 = convertChannel(token.value);
+          input.consume();
```

Comment:
Am I wrong or does this imply that there's a channel name on the right hand side of the expression? Can you do that? Can you even say something like `GEM_SHOT_NUM_VALUE` `=` `Shot Number`?

Reply:
Yes, in eCat you can compare columns to each other and this is in the requirements - but it looks like this isn't possible with the API right now. Is this a necessary feature? If so then @MRichards99 and @kevinphippsstfc need to know about it, if we don't know then we need to ask @mattjSTFC

Reply:
Following a discussion on Slack, this will be tracked in a Jira issue once we have clarification from CLF